eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,521 @@
+	Opcode       vm.OpCode `json:"opcode"`
+}
+
+type callFrameWithOpcodes struct {
+	Type             vm.OpCode       `json:"-"`
+	From             common.Address  `json:"from"`
+	Gas              uint64          `json:"gas"`
+	GasUsed          uint64          `json:"gasUsed"`
+	To               *common.Address `json:"to,omitempty" rlp:"optional"`
+	Input            []byte          `json:"input" rlp:"optional"`
+	Output           []byte          `json:"output,omitempty" rlp:"optional"`
+	Error            string          `json:"error,omitempty" rlp:"optional"`
+	RevertReason     string          `json:"revertReason,omitempty"`
+	Logs             []callLog       `json:"logs,omitempty" rlp:"optional"`
+	Value            *big.Int        `json:"value,omitempty" rlp:"optional"`
+	revertedSnapshot bool
+
+	AccessedSlots     accessedSlots                              `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address                           `json:"extCodeAccessInfo"`
+	UsedOpcodes       map[vm.OpCode]uint64                       `json:"usedOpcodes"`
+	ContractSize      map[common.Address]*contractSizeWithOpcode `json:"contractSize"`
+	OutOfGas          bool                                       `json:"outOfGas"`
+	Calls             []callFrameWithOpcodes                     `json:"calls,omitempty" rlp:"optional"`
+}
+
+func (f callFrameWithOpcodes) TypeString() string {
+	return f.Type.String()
+}
+
+func (f callFrameWithOpcodes) failed() bool {
+	return len(f.Error) > 0 && f.revertedSnapshot
+}
+
+func (f *callFrameWithOpcodes) processOutput(output []byte, err error, reverted bool) {
+	output = common.CopyBytes(output)
+	// Clear error if tx wasn't reverted. This happened
+	// for pre-homestead contract storage OOG.
+	if err != nil && !reverted {
+		err = nil
+	}
+	if err == nil {
+		f.Output = output
+		return
+	}
+	f.Error = err.Error()
+	f.revertedSnapshot = reverted
+	if f.Type == vm.CREATE || f.Type == vm.CREATE2 {
+		f.To = nil
+	}
+	if !errors.Is(err, vm.ErrExecutionReverted) || len(output) == 0 {
+		return
+	}
+	f.Output = output
+	if len(output) < 4 {
+		return
+	}
+	if unpacked, err := abi.UnpackRevert(output); err == nil {
+		f.RevertReason = unpacked
+	}
+}
+
+type callFrameWithOpcodesMarshaling struct {
+	TypeString        string `json:"type"`
+	Gas               hexutil.Uint64
+	GasUsed           hexutil.Uint64
+	Value             *hexutil.Big
+	Input             hexutil.Bytes
+	Output            hexutil.Bytes
+	AccessedSlots     accessedSlots                              `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address                           `json:"extCodeAccessInfo"`
+	UsedOpcodes       map[vm.OpCode]uint64                       `json:"usedOpcodes"`
+	ContractSize      map[common.Address]*contractSizeWithOpcode `json:"contractSize"`
+	OutOfGas          bool                                       `json:"outOfGas"`
+	Calls             []callFrameWithOpcodes                     `json:"calls,omitempty"`
```

Comment:
In the Marshaling struct you only need to specify the fields which you wish to override the type for. So in this case `Calls` is not necessary here, Same with OutOfGas, ContractSize, ExtCodeAccessInfo and AccessedSlots.

The reason for using this double-struct approach is so that in the tracer logic we can simply use normal Go types (like uint64) and when calling json.Marshal it will "automatically" understand how to encode them because we provide hints here that encoding should use these other specialized types (e.g. hexutil.Uint64).